### PR TITLE
ADTs, Control flow and Generics sections for preliminaries

### PR DESCRIPTION
**Pull request by pvdrz**
This PR contains:
- A section about algebraic data types.
- A secition about control flow.
- A section about generics.
Code examples for each section are needed but the text is ready. After merging this PR the preliminaries will be ready.

### CODE REVIEW

File: document/chapters/preliminaries.tex [hunk 1]
```diff
@@ -118,8 +118,23 @@ \subsection{Lifetimes}
 In principle, every reference needs a lifetime annotation. However, this is avoided by a process known as lifetime elision, where lifetimes are inferred automatically by the compiler. Even then, in certain cases the programmer might need to add such annotations. However, this will be discussed in the generics section.
 
 \section{Algebraic data types}
+Rust has both sum and product algebraic data types in the form of structures and enumerations respectively.
```
Comment (ncardozo) [inline document/chapters/preliminaries.tex]:
```suggestion
Rust has both sum and product algebraic data types~\cite{REF} in the form of structures and enumerations respectively.
```

File: document/chapters/preliminaries.tex [hunk 2]
```diff
@@ -118,8 +118,23 @@ \subsection{Lifetimes}
...
+
+Structures are named sum types, where an struct type has a fixed set of named fields. When declaring a new value of an struct type, all its field must be given as shown in TODO. Enumerations are named product types, where an enum type has a fixed set of named variants. When declaring a new value of an enum type, one single variant must be chosen as shown in TODO.
```
Comment (ncardozo) [inline document/chapters/preliminaries.tex]:
```suggestion
Structures are named sum types, where an struct type has a fixed set of named fields. When declaring a new value of an struct type, all its fields must be given as shown in TODO. Enumerations are named product types, where an enum type has a fixed set of named variants. When declaring a new value of an enum type, one single variant must be chosen as shown in TODO.
```

File: document/chapters/preliminaries.tex [hunk 3]
```diff
@@ -118,8 +118,23 @@ \subsection{Lifetimes}
...
+
+It is also possible to add associated functions to any type using the \texttt{impl} keyword in an object oriented programming style. Each associated function could or could not use an instance of a given type, just as instance and static methods in languages such as Java.
```
Comment (ncardozo) [inline document/chapters/preliminaries.tex]:
```suggestion
It is also possible to add associated functions to any type using the \texttt{impl} keyword in an object oriented programming style. Each associated function could or could not use an instance of a given type. This is similar, for example, to the way instance and static methods are defined in Java.
```

File: document/chapters/preliminaries.tex [hunk 4]
```diff
@@ -118,8 +118,23 @@ \subsection{Lifetimes}
...
+
 \section{Control flow}
+Rust has both the usual \texttt{if/else} conditional and \texttt{while} loop statements as shown in TODO and TODO respectively. \texttt{for} loops are iterator based, where the variable to be iterated must implement the \texttt{Iterator} trait.
```
Comment (ncardozo) [inline document/chapters/preliminaries.tex]:
```suggestion
Control flow in Rust is realized using the common \texttt{if/else} conditional and \texttt{while} loop statements, as in other programming languages. \texttt{for} loops are iterator based, where the variable to be iterated must implement the \texttt{Iterator} trait.
```

File: document/chapters/preliminaries.tex [hunk 5]
```diff
@@ -118,8 +118,23 @@ \subsection{Lifetimes}
...
+
+Rust has pattern matching capabilities thanks to its functional heritage, pattern matching in Rust can be used to match specific values or to destructure tuples, arrays, enumerations or structures. There are three statements to do pattern matching: \texttt{match}, \texttt{if let} and \texttt{while let}. Examples for these statemens can be found in TODO, TODO and TODO respectively. 
 \section{Generics}
+From an external perspective, generic types in Rust are quite similar to the generic types in Java, i.e. every type can have generic type parameters in order to reduce code duplication and such parameters are specified between angled brackets. However, there are three main differences:
```
Comment (ncardozo) [inline document/chapters/preliminaries.tex]:
```suggestion
From an external perspective, generic types in Rust are quite similar to the generic types in Java --that is, every type can have generic type parameters in order to reduce code duplication and such parameters are specified between angled brackets. However, there are three main differences between the two implementations:
```

File: document/chapters/preliminaries.tex [hunk 6]
```diff
@@ -118,8 +118,23 @@ \subsection{Lifetimes}
...
+\begin{itemize}
+    \item Rust allows lifetimes as generic parameters. These are used when the borrow checker cannot infer a proper lifetime for a variable, and thus explicit lifetime annotations are required. An example of this can be found in TODO.
+    \item In Java, generic parameters can be bounded by forcing them to be instances of a class or interface. In Rust, which is not an object oriented language, bounds are done over traits instead, as in TODO.
+    \item Internally, generics in Java are implemented doing type erasure, where each generic parameter bounds are checked, and then the Java compiler forgets about the type of such paremeter and uses dynamic dispatch over the type \texttt{Object}. On the other hand, Rust uses monomorphization, where the compiler builds a new type specialized for each instance of a generic parameter making all the dispatch completely static, as shown in TODO. \footnote{Rust allows type erasure via trait objects. However, this goes is beyond the scope of this work.}
```
Comment (ncardozo) [inline document/chapters/preliminaries.tex]:
```suggestion
    \item Internally, generics in Java are implemented doing type erasure, where each generic parameter bounds are checked, and then the Java compiler forgets about the type of such paremeter and uses dynamic dispatch over the type \texttt{Object}. On the other hand, Rust uses monomorphization, where the compiler builds a new type specialized for each instance of a generic parameter making all the dispatch completely static, as shown in TODO.\footnote{Rust allows type erasure via trait objects. However, this goes is beyond the scope of this work.}
```

**Review by ncardozo — CHANGES_REQUESTED** [review]

File: document/chapters/preliminaries.tex (hunk 1 above)
Comment (pvdrz) [reply document/chapters/preliminaries.tex]: We need a reference about the notion of algebraic data types? or about the fact that Rust has algebraic data types?

**Comment by pvdrz** [comment]
I believe this chapter is almost ready. We need to address the citation for ADTs, which paper/book should we cite there?

**Pull request merged**